### PR TITLE
Failpoints

### PR DESCRIPTION
**Pull request by knizhnik**
See https://discord.com/channels/869525774699462656/922518319637090354

### CODE REVIEW

File: control_plane/src/local_env.rs [hunk 1]
```diff
@@ -72,6 +72,18 @@ pub struct PageServerConf {
...
+    pub gc_period: String,
+
+    // GC horizon
+    pub gc_horizon: u64,
+
+    // Checkpoint period
+    pub checkpoint_period: String,
+
+    // Checkpoint distance
+    pub checkpoint_distance: u64,
```
Comment (funbringer) [inline control_plane/src/local_env.rs:86]: Looks like this change is unrelated to failpoints. Could you extract this into a separate patch?

File: pageserver/src/bin/pageserver.rs [hunk 2]
```diff
@@ -488,6 +490,13 @@ fn main() -> Result<()> {
     // as a ref.
     let conf: &'static PageServerConf = Box::leak(Box::new(conf));
 
+    let scenario = FailScenario::setup();
+    if fail::has_failpoints() {
+        std::panic::set_hook(Box::new(|_| {
+            std::process::exit(1);
+        }));
+    }
+
```
Comment (lubennikovaav) [inline pageserver/src/bin/pageserver.rs:499]: I don't quite understand the purpose of this code.
`FailScenario::setup()` configures fail points specified in the `FAILPOINTS` environment variable. 
As far as I see it is never used in the tests.

Have you set it manually?

File: test_runner/fixtures/zenith_fixtures.py [hunk 3]
```diff
@@ -431,6 +431,10 @@ def __init__(self, config: ZenithEnvBuilder):
 pg_port = {pageserver_port.pg}
 http_port = {pageserver_port.http}
 auth_type = '{pageserver_auth_type}'
+gc_period = '1s'
+gc_horizon = 67108864
+checkpoint_period = '1s'
+checkpoint_distance = 1048576
```
Comment (lubennikovaav) [inline test_runner/fixtures/zenith_fixtures.py:437]: Using non-default settings in all tests doesn't look right to me.
Let's tune them for one test, if necessary. Or change defaults if we believe that they are bad.

File: test_runner/batch_others/test_recovery.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                    # Sleep for some time to let checkpoint create image layers
+                    time.sleep(2)
+
+                    # Configure failpoints
+                    pscur.execute(
+                        "failpoints checkpoint-before-sync=sleep(2000);checkpoint-after-sync=panic")
+
+                    # Do some updates until pageserver is crashed
+                    try:
+                        while True:
```
Comment (lubennikovaav) [inline test_runner/batch_others/test_recovery.py:39]: Maybe add some timeout in case checkpointer fails to fail?

File: control_plane/src/local_env.rs [hunk 5]
```diff
@@ -81,6 +93,10 @@ impl Default for PageServerConf {
             http_port: 0,
             auth_type: AuthType::Trust,
             auth_token: "".to_string(),
+            gc_period: "100s".to_string(),
+            gc_horizon: 64 * 1024 * 1024,
+            checkpoint_period: "1s".to_string(),
+            checkpoint_distance: 256 * 1024 * 1024,
```
Comment (lubennikovaav) [inline control_plane/src/local_env.rs:99]: Why not use DEFAULT_CHECKPOINT_PERIOD and other `defaults` here?

File: .circleci/config.yml [hunk 6]
```diff
@@ -126,7 +126,7 @@ jobs:
             fi
 
             export CARGO_INCREMENTAL=0
-            "${cov_prefix[@]}" cargo build $CARGO_FLAGS --bins --tests
+            "${cov_prefix[@]}" cargo build $CARGO_FLAGS --features fail/failpoints --bins --tests
```
Comment (lubennikovaav) [inline .circleci/config.yml:129]: Do we need a similar change for `cargo build` in Makefile?

**Review by lubennikovaav — CHANGES_REQUESTED** [review]

File: pageserver/src/bin/pageserver.rs (hunk 2 above)
Comment (funbringer) [reply pageserver/src/bin/pageserver.rs:499]: I suspect @knizhnik wanted to make it testable/usable in manual mode while developing this feature. Either way, I'd leave it be since it's useful.

File: test_runner/fixtures/zenith_fixtures.py (hunk 3 above)
Comment (funbringer) [reply test_runner/fixtures/zenith_fixtures.py:437]: I'd argue the change doesn't belong to this PR.

Comment (knizhnik) [reply test_runner/fixtures/zenith_fixtures.py:437]: This changes are needed for test_recovery.py test included in this PR.

File: pageserver/src/bin/pageserver.rs (hunk 2 above)
Comment (knizhnik) [reply pageserver/src/bin/pageserver.rs:499]: The purpose of this code is to cause pageserver termination in case of panic. `failpoint` module doesn't provide some builtin action to terminate process. It can only thrown panic which terminates just current thread. And if we want to test abnormal process termination we need to terminate the whole process.

File: control_plane/src/local_env.rs (hunk 1 above)
Comment (knizhnik) [reply control_plane/src/local_env.rs:86]: This changes are needed for test_recovery.py test included in this PR.

File: test_runner/batch_others/test_recovery.py (hunk 4 above)
Comment (knizhnik) [reply test_runner/batch_others/test_recovery.py:39]: It seems to be no normal reasons for pageserver to "fail to fail". If it happen, then it is bug as hanging of any other test. There is CI timeout to terminate test run in this case. Do we need some extra timeout in each test? Certainly if this test hangs and CI run is terminated by timeout, then we will not get results for other tests which are not launched. If it will happen frequently, then it may be a problem and we really need to somehow explicitly prevent hanging in  test_recovery test. By right now I prefer not to solve this problem.

File: control_plane/src/local_env.rs (hunk 5 above)
Comment (knizhnik) [reply control_plane/src/local_env.rs:99]: 1 sec and 256Mb seems to be default checkpoint values.
I just prefer to specify them explicitly in this file, because setting for tests may be different with one used in productions. So not want that checkpoint interval for tests depends on default setting for production.

File: .circleci/config.yml (hunk 6 above)
Comment (knizhnik) [reply .circleci/config.yml:129]: I do not want failpoints to be linked in production. It should be used only for tests. When we are doing "cargo build" there is no information whether we are doing it to run tests or we are building images for production. Certainly we can create special make target for building zenith with safepoints. For me  manual specification of `--features fail/failpoints` works fine. Is there any other way except creating special make target for  builds with safepoints?

Including safepoints only in debug builds IMHO is not also good option, because debug version is soooo slow.

File: control_plane/src/local_env.rs (hunk 1 above)
Comment (funbringer) [reply control_plane/src/local_env.rs:86]: I'm trying to say that IMO failpoints infra patch should include neither the test nor those settings.

Comment (knizhnik) [reply control_plane/src/local_env.rs:86]: IMHO it is strange to commit some feature without test.

**Comment by lubennikovaav** [comment]
I've opened a new PR with rebased version of this one: #1324
